packages/vite/src/node/ssr/ssrTransform.ts

```diff
@@ -59,124 +53,159 @@ export async function ssrTransform(
     )
   }
 
-  // 1. check all import statements and record id -> importName map
-  for (const node of ast.body as Node[]) {
-    // import foo from 'foo' --> foo -> __import_foo__.default
-    // import { baz } from 'foo' --> baz -> __import_foo__.baz
-    // import * as ok from 'foo' --> ok -> __import_foo__
-    if (node.type === 'ImportDeclaration') {
-      const importId = defineImport(node, node.source.value as string)
-      for (const spec of node.specifiers) {
-        if (spec.type === 'ImportSpecifier') {
-          idToImportMap.set(
-            spec.local.name,
-            `${importId}.${spec.imported.name}`
-          )
-        } else if (spec.type === 'ImportDefaultSpecifier') {
-          idToImportMap.set(spec.local.name, `${importId}.default`)
-        } else {
-          // namespace specifier
-          idToImportMap.set(spec.local.name, importId)
-        }
+  const cjsRE = /\b(?:(module\.)?(exports)(\.\w+)?\b *(=)?|(require)\()/g
+  let match: RegExpExecArray | null
+  let isCommonJS = false
+  while ((match = cjsRE.exec(code))) {
+    isCommonJS = true
+    if (match[1]) {
+      const moduleExportsEnd = match.index + 'module.exports'.length
+      s.overwrite(match.index, moduleExportsEnd, ssrModuleExportsKey)
+      if (!match[3] && match[4]) {
+        s.appendLeft(moduleExportsEnd, '.default')
       }
-      s.remove(node.start, node.end)
+    } else if (match[2]) {
+      s.overwrite(
+        match.index,
+        match.index + 'exports'.length,
+        ssrModuleExportsKey
+      )
+    } else if (match[5]) {
+      s.overwrite(match.index, match.index + 'require'.length, ssrImportKey)
     }
   }
 
-  // 2. check all export statements and define exports
-  for (const node of ast.body as Node[]) {
-    // named exports
-    if (node.type === 'ExportNamedDeclaration') {
-      if (node.declaration) {
-        if (
-          node.declaration.type === 'FunctionDeclaration' ||
-          node.declaration.type === 'ClassDeclaration'
-        ) {
-          // export function foo() {}
-          defineExport(node.declaration.id!.name)
-        } else {
-          // export const foo = 1, bar = 2
-          for (const declaration of node.declaration.declarations) {
-            const names = extractNames(declaration.id as any)
-            for (const name of names) {
-              defineExport(name)
-            }
-          }
-        }
-        s.remove(node.start, (node.declaration as Node).start)
-      } else if (node.source) {
-        // export { foo, bar } from './foo'
+  if (!isCommonJS) {
+    const ast = parser.parse(code, {
+      sourceType: 'module',
+      ecmaVersion: 2021,
+      locations: true
+    }) as any
+
+    // mark this as a module
+    s.prepend(
+      `Object.defineProperty(${ssrModuleExportsKey}, "__esModule", {value: true})\n`
```

Comment:
Injecting `__esModule` definition here (instead of defining it in `instantiateModule`) allows us to make CommonJS modules compatible with ESM imports by doing `exports.default = exports` for them when undefined.